app/services/optout/OptOutService.scala

```diff
@@ -93,8 +97,77 @@ class OptOutService @Inject()(itsaStatusUpdateConnector: ITSAStatusUpdateConnect
+    val result = Future.sequence(responsesSeqOfFutures).
       map(responsesSeq => findAnyFailOrFirstSuccess(responsesSeq))
+
+//    def checkVoluntaryElseReturnCurrent(optOutTaxYear: OptOutTaxYear): String = {
+//
+//      val isYearUserWantsToUpdate = intentTaxYear == optOutTaxYear.taxYear
+//
+//      if (isYearUserWantsToUpdate){
+//        optOutTaxYear.status match {
+//          case Voluntary => s"Voluntary change :: Successful"
+//          case Annual => s"Annual change :: Failure"
+//          case Mandated => s"Mandated change :: Failure"
+//          case _ => s"Other non-changeable status :: Failure"
+//        }
+//      }else{
+//        s"Status of ${optOutTaxYear.taxYear} : ${optOutTaxYear.status} remained the same"
+//      }
+//    }
+
+    val x = CheckYourAnswersAuditModel(
+      nino = user.nino,
+      optOutRequestedFromTaxYear = intentTaxYear.formatTaxYearRange,
+      currentYear = optOutProposition.currentTaxYear.taxYear.toString,
+
+      beforeITSAStatusCurrentYearMinusOne = optOutProposition.previousTaxYear.status.toString,
+      beforeITSAStatusCurrentYear = optOutProposition.currentTaxYear.status.toString,
+      beforeITSAStatusCurrentYearPlusOne = optOutProposition.nextTaxYear.status.toString,
+
+      outcome = createOutcome(result),
+
+      afterAssumedITSAStatusCurrentYearMinusOne = optOutProposition.previousTaxYear.expectedItsaStatusAfter(intentTaxYear).toString,
+      afterAssumedITSAStatusCurrentYear = optOutProposition.currentTaxYear.expectedItsaStatusAfter(intentTaxYear).toString,
+      afterAssumedITSAStatusCurrentYearPlusOne = optOutProposition.nextTaxYear.expectedItsaStatusAfter(intentTaxYear).toString,
+
+      currentYearMinusOneCrystallised = optOutProposition.previousTaxYear.crystallised
+    )
+    println(s"!!!!!!!!!!!!!!! Auditing Model:" +
+      s"\nnino: ${x.nino} " +
+      s"\noptOutRequestedFromTaxYear: ${x.optOutRequestedFromTaxYear} " +
+      s"\ncurrentYear: ${x.currentYear} " +
+      s"\nbeforeITSAStatusCurrentYearMinusOne: ${x.beforeITSAStatusCurrentYearMinusOne} " +
+      s"\nbeforeITSAStatusCurrentYear: ${x.beforeITSAStatusCurrentYear} " +
+      s"\nbeforeITSAStatusCurrentYearPlusOne: ${x.beforeITSAStatusCurrentYearPlusOne} " +
+      s"\noutcome: ${x.outcome} " +
+      s"\nafterAssumedITSAStatusCurrentYearMinusOne: ${x.afterAssumedITSAStatusCurrentYearMinusOne} " +
+      s"\nafterAssumedITSAStatusCurrentYear: ${x.afterAssumedITSAStatusCurrentYear} " +
+      s"\nafterAssumedITSAStatusCurrentYearPlusOne: ${x.afterAssumedITSAStatusCurrentYearPlusOne} " +
+      s"\ncurrentYearMinusOneCrystallised: ${x.currentYearMinusOneCrystallised}")
+
+    auditingService.extendedAudit(x)
+
+    result
+  }
+  private def createOutcome(futureResponse: Future[ITSAStatusUpdateResponse]): Outcome = {
+
+    val resolved = futureResponse.value.getOrElse(ITSAStatusUpdateResponseFailure)
+    resolved match {
+      case ITSAStatusUpdateResponseFailure => new Outcome {
```

Comment:
Why is Outcome not just a case class with fields: isSuccessful, failureReason and failureCategory?

Reply:
If we can avoid usage of "getOrElse" over Future as this make hard to understand / reason outcome of this operation;

Reply:
@simondrugan16 I'm not sure what you mean, currently Outcome is a case class with those fields.
@avelx thanks to @kabuumu these issues should now be resolved

Reply:
Ah I see. When you have default values (like you have in Outcome) you can still construct the model normally, e.g: Outcome(false, "reason", "whateverElse"). I think it would be best to remove the default params in fact since there's no reason to have them, plus it's prone to breaking, for example if someone constructs an Outcome like Outcome(), they will actually have an Outcome(isSuccessful = true, failureCategory = "API_FAILURE", failureReason = "Failure reasons") which is not correct in any scenario. Also in the audit model when you had a success then you would still have the fields failureCategory and failureReason which would have an empty string in them. It would be best if these just disappeared when empty imo (this is how we do TxM events elsewhere too.

A solution to fix this would be to remove all the default arguments, then make the failure fields optional. You can use the ++ syntax (there are examples in other TxM events) which will remove the fields from the json object if they are None. Done like this you can construct the Outcome objects without any overrides too. I'm happy to help if you have questions about this

Reply:
@simondrugan16 
Alrighty, so locally I've changed Outcome, to your suggestions, I'm a bit unsure of how the ++ syntax works with JSON and leaving the fields out if it's Null 🤔 
If you have any examples or docs on it that would be greatly appreciated, or if you're up for a call to go over it maybe?